python/ray/serve/backend_state.py

```diff
@@ -596,22 +621,35 @@ def delete_backend(self, backend_tag: BackendTag,
         return new_goal_id
 
     def _stop_wrong_version_replicas(
-            self, backend_tag: BackendTag, version: str,
-            graceful_shutdown_timeout_s: float) -> int:
+            self, backend_tag: BackendTag, target_replicas: int,
+            target_version: str, graceful_shutdown_timeout_s: float) -> int:
         # NOTE(edoakes): this short-circuits when using the legacy
         # `create_backend` codepath -- it can be removed once we deprecate
         # that as the version should never be None.
-        if version is None:
+        if target_version is None:
             return
 
-        # TODO(edoakes): to implement rolling upgrade, all we should need to
-        # do is cap the number of old version replicas that are stopped here.
-        replicas_to_stop = self._replicas[backend_tag].pop(
-            exclude_version=version,
+        old_running_replicas = self._replicas[backend_tag].count(
+            exclude_version=target_version,
             states=[
                 ReplicaState.SHOULD_START, ReplicaState.STARTING,
                 ReplicaState.RUNNING
             ])
+        new_running_replicas = self._replicas[backend_tag].count(
+            version=target_version, states=[ReplicaState.RUNNING])
+
+        pending_replicas = (
+            target_replicas - new_running_replicas - old_running_replicas)
+        rollout_size = max(int(0.2 * target_replicas), 1)
```

Comment:
Would be good to add a comment saying that rollout_size is the max number of replicas we ever want to be not running any version